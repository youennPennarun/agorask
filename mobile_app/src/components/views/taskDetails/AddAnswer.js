/* @flow */

import React, {Component, PropTypes} from 'react';

import {connect} from 'react-redux';
import {login} from '../../../redux/actions/router';

import {View, Text, StyleSheet} from 'react-native';
import {MKButton, MKTextField} from 'react-native-material-kit';

type AddAnswerStateType = {
  answerHeight: number,
  loading: boolean,
  answer: string,
};

type AddAnswerPropsType = {
  login: Function,
  addAnswer: Function,
  token: string,
};

export class AddAnswer extends Component {
  static propTypes = {
    login: PropTypes.func.isRequired,
    addAnswer: PropTypes.func.isRequired,
    token: PropTypes.string,
  }
  state: AddAnswerStateType = {
    answerHeight: 0,
    loading: false,
    answer: '',
  };

  props: AddAnswerPropsType;

  updateAnswerFieldHeight({height}) {
  // height: number
    if (this.state.answerHeight !== height) {
      this.setState({answerHeight: height});
    }
  }


  submit() {
    const {answer} = this.state;
    this.setState({loading: true});
    this.props.addAnswer({answer: answer}, this.props.token)
      .then(() => {
          this.setState({loading: false, answer: ''});
      })
      .catch(e => {
        console.log('error ', e);
        this.setState({loading: false});
      });
  }

  _renderNotLoggedin(): React.Element<*> {
    return (
      <View style={styles.notLoggedInContainer}>
        <Text>You need to be logged in to add an answer</Text>
        <LoginBtn onPress={() => { this.props.login(); }} />
      </View>
    );
  }

  render(): React.Element<*> {
    if (!this.props.token) return this._renderNotLoggedin();
    return (
      <View style={styles.container} >
        <MKTextField style={[styles.answerInput, {height: Math.max(35, this.state.answerHeight)}]}
          multiline
          value={this.state.answer}
          onTextChange={text => { this.setState({answer: text}); }}
          onChange={({nativeEvent: {contentSize}}) => {
            this.updateAnswerFieldHeight(contentSize);
          }}
          onContentSizeChange={({nativeEvent: {contentSize}}) => {
            this.updateAnswerFieldHeight(contentSize);
          }}
          placeholder='answer' />

        <AddAnswerBtn enabled={!this.state.loading} onPress={() => { this.submit(); }} />
      </View>

    );
  }
}

const styles = StyleSheet.create({
  container: {
    marginHorizontal: 30,
    paddingVertical: 10,
    marginBottom: 25,
  },
  answerInput: {
  },
  button: {
    alignSelf: 'center',
    width: 150,
    marginVertical: 10,
  },
  notLoggedInContainer: {
    alignItems: 'center',
  },
});

const AddAnswerBtn = MKButton.coloredButton()
  .withText('Add an answer')
  .withStyle(styles.button)
  .build();
const LoginBtn = MKButton.coloredButton()
  .withText('Login')
  .withStyle(styles.button)
  .build();

function mapStateToProps(state): Object {
  return {
    ...state.user,
  };
}
function mapDispatchToProps(dispatch): Object {
  return {
    login: () => { dispatch(login()); },
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(AddAnswer);
