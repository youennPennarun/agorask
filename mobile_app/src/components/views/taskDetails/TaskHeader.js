/* @flow */

import React, {PropTypes} from 'react';
import {
  View,
  Animated,
  Text,
  StyleSheet,
} from 'react-native';
import gql from 'graphql-tag';
import ProfilePic from '../../commons/ProfilePic';

import moment from 'moment';

type TaskHeaderStateType = {
  radius: Animated.Value,
  marginHorizontal: Animated.Value,
  blockDateTranslationY: Animated.Value,
  dateOpacity: Animated.Value,
  nbAnswersOpacity: Animated.Value,
}

type TaskHeaderPropsType = {
  title: string,
  date: ?Date,
  postedBy: {
    username: string,
  },
  nbAnswers: number,
}

const initialState = {
  radius: 3,
  marginHorizontal: 10,
  blockDateTranslationY: 15,
  dateOpacity: 0,
  nbAnswersOpacity: 1,
};

class TaskHeader extends React.Component {

  static propTypes = {
    title: PropTypes.string.isRequired,
    nbAnswers: PropTypes.number.isRequired,
    postedBy: PropTypes.shape({
      username: PropTypes.string.isRequired,
    }).isRequired,
  };

  state: TaskHeaderStateType = {
    radius: new Animated.Value(initialState.radius),
    marginHorizontal: new Animated.Value(initialState.marginHorizontal),
    blockDateTranslationY: new Animated.Value(initialState.blockDateTranslationY),
    dateOpacity: new Animated.Value(initialState.dateOpacity),
    nbAnswersOpacity: new Animated.Value(initialState.nbAnswersOpacity),
  }

  props: TaskHeaderPropsType;

  componentDidMount() {
    const {
      radius,
      marginHorizontal,
      blockDateTranslationY,
      dateOpacity,
      nbAnswersOpacity,
    } = this.state;

    if (this.props.full) {
      Animated.parallel([
        Animated.timing(marginHorizontal, {
          toValue: 0,
          duration: 1000,
        }),
        Animated.timing(radius, {
          toValue: 0,
          duration: 1000,
        }),

        Animated.timing(blockDateTranslationY, {
          toValue: 0,
          useNativeDriver: true,
        }),
      ]).start();
    }
  }
  componentWillReceiveProps(next: TaskHeaderPropsType) {
    const {
      radius,
      marginHorizontal,
      blockDateTranslationY,
    } = this.state;
    if (this.props.full !== next.full && !next.full) {
      Animated.parallel([
        Animated.timing(blockDateTranslationY, {
          toValue: initialState.blockDateTranslationY,
          useNativeDriver: true,
        }),
        Animated.timing(radius, {
          toValue: initialState.radius,
        }),
        Animated.timing(marginHorizontal, {
          toValue: initialState.marginHorizontal,
        }),
      ]).start();
    }
  }
  render(): React.Element<*> {
    const {title, date, postedBy} = this.props;
    return (
      <Animated.View style={[
        styles.questionContainer,
        {
          elevation: (date) ? 3 : 1,
          marginHorizontal: this.state.marginHorizontal,
          borderRadius: this.state.radius,
        }]} >
        <ProfilePic style={styles.pic} size={50} username={this.props.postedBy.username} />
        <View style={styles.leftColumn}>
          <Text style={styles.question} numberOfLines={2} ellipsizeMode='tail' >{title}</Text>
          <Text style={styles.postedBy} >
            {'Posted by: '}
            <Text style={styles.postedByUsername}>
              {postedBy.username}
            </Text>
          </Text>
        </View>
        <View style={styles.rightColumn}>
          <View style={styles.blockDate} >
            <Text>{moment(date).format('DD/MM/YY')}</Text>
          </View>
          <View style={styles.blockNbAnswers} >
            <Text>28 answers</Text>
          </View>
        </View>
      </Animated.View>
    );
  }
}

const styles = StyleSheet.create({
  questionContainer: {
    flexDirection: 'row',
    height: 80,
    backgroundColor: 'white',
    elevation: 1,
    marginBottom: 0,
    paddingLeft: 5,
  },
  pic: {
    alignSelf: 'center',
  },
  leftColumn: {
    flex: 0.70,
    alignItems: 'flex-start',
    justifyContent: 'center',
  },
  question: {
    marginLeft: 5,
    textAlignVertical: 'center',
    fontSize: 16,
    fontWeight: '800',
    color: '#212121',
  },
  postedBy: {
    fontWeight: '100',
    fontSize: 12,
    marginLeft: 15,
  },
  postedByUsername: {
    color: '#020202',
    fontSize: 13,
  },
  rightColumn: {
    flex: 0.30,
    marginRight: 5,
    marginVertical: 5,
    alignItems: 'flex-end',
    justifyContent: 'space-between',
  },
  blockDate: {
    alignItems: 'flex-start',
  },
  blockNbAnswers: {
    alignItems: 'flex-end',
  },
});

TaskHeader.fragments = {
  task: gql`
    fragment Task on Task {
      _id
      title,
      nbAnswers,
      postedBy {
        username,
      }
    }
  `,
};

export default TaskHeader;
