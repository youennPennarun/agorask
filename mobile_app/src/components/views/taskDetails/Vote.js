import React, { PropTypes} from 'react';
import {View, Text, StyleSheet} from 'react-native';

export default function Vote(props): any {
  const {score} = props;

  let scoreStatusStyle = {};
  if (score > 1) {
    scoreStatusStyle = styles.good;
  } else if (score < -1) {
    scoreStatusStyle = styles.bad;
  }
  return (
    <View style={styles.container}>
      <Text style={styles.arrow}>+</Text>
      <Text style={[styles.score, scoreStatusStyle]}>{score}</Text>
      <Text style={styles.arrow}>-</Text>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    width: 55,
    height: 80,
    justifyContent: 'space-between',
    alignItems: 'center',
    paddingTop: 5,
    paddingBottom: 5,
  },
  arrow: {
    flex: 1.1,
    fontSize: 19,
  },
  score: {
    fontSize: 17,
    flex: 0.8,
  },
  good: {
    color: 'green',
  },
  bad: {
    color: 'red',
  },
});

Vote.propTypes = {
  score: PropTypes.number.isRequired,
};
