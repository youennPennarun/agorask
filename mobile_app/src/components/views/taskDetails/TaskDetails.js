import React, {Component, PropTypes} from 'react';
import {
  View,
  Animated,
  Text,
  StyleSheet,
  Dimensions,
  ScrollView,
  InteractionManager,
  BackAndroid,
} from 'react-native';
import {
  MKProgress,
  MKSpinner,
} from 'react-native-material-kit';
import update from 'immutability-helper';
import {connect} from 'react-redux';


import { graphql, compose } from 'react-apollo';
import gql from 'graphql-tag';
import moment from 'moment';
import {login} from '../../../redux/actions/router';

import TaskHeader from './TaskHeader';
import Vote from './Vote';
import AddAnswer from './AddAnswer';


const {width, height} = Dimensions.get('window');

type TaskDetailsPropsType = {
  id: string,
  task: {
    name: string,
    title: string,
    date: string,
    answers: Array,
    postedBy: {
      username: string,
    },
  },
}

type TaskDetailsStateType = {
  bodyHeight: Animated.Value,
  isHiding: boolean,
};

export class TaskDetails extends Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
    task: PropTypes.shape({
      name: PropTypes.string,
      title: PropTypes.string.isRequired,
      answers: PropTypes.array,
      postedBy: PropTypes.object,
    }).isRequired,
  };
  static defaultProps = {
    task: {
      postedBy: {
        username: '',
      },
    },
  };
  state: TaskDetailsStateType = {
    bodyHeight: new Animated.Value(0),
    isHiding: false,
  };
  props: TaskDetailsPropsType;

  componentDidMount() {
    BackAndroid.addEventListener('hardwareBackPress', this.onBack);
    Animated.timing(this.state.bodyHeight, {
      toValue: height,
      duration: 200,
    }).start();
  }
  componentWillUnmount() {
    BackAndroid.removeEventListener('hardwareBackPress', this.onBack);
  }

  onBack = (): boolean => {
    this.setState({isHiding: true});
    Animated.timing(this.state.bodyHeight, {
      toValue: 0,
      duration: 200,
    }).start();
    return false;
  }

  addAnswer(answer: string): Promise<Object> {
    return this.props.addAnswer(this.props.task._id, answer, this.props.token);
  }

  _renderAnswer({_id, answer, postedBy: {username}, date, rating = 0, userRating}, key): any {
    const {token, rateAnswer, task} = this.props;
    const onVoteCallback = (token) ? (value => {
      rateAnswer(task._id, _id, value, token);
    }) : undefined;
    return (
      <View key={key} style={styles.answerContainer}>
        <View style={styles.row} >
          <Vote score={rating} userRating={userRating} onVote={onVoteCallback} />
          <View style={styles.answer} >
            <Text style={styles.answerText}>{answer}</Text>
            <View style={styles.answerFooter} >
              <Text style={styles.username} >{username}</Text>
              <Text style={styles.date} >{moment(date).format('DD/MM/YY')}</Text>
            </View>
          </View>
        </View>
        <View style={styles.separator} />
      </View>
    );
  }
  _renderProgressBar(): ?MKProgress.Indeterminate {
    const {loading, task} = this.props;
    const {answers = []} = task;
    if (!loading || !answers.length) return null;
    return <MKProgress.Indeterminate style={styles.progress} />;
  }
  _renderSpinner(): ?MKSpinner {
    const {loading, task} = this.props;
    const {answers = []} = task;
    if (!loading || answers.length) return null;
    return <MKSpinner style={styles.spinner} />;
  }
  render(): React.Element {
    console.log('{{{{{{{{{{{{{{{{{{{{{{{{{')
    console.log(this.props.task)
    console.log('{{{{{{{{{{{{{{{{{{{{{{{{{')
    const {title, date, answers = [], postedBy = {username: ''}} = this.props.task;
    return (
      <View style={styles.container} >
        {this._renderProgressBar()}
        <ScrollView>
          <TaskHeader title={title}
            full={(!this.state.isHiding)}
            postedBy={postedBy}
            date={date}
            nbAnswers={answers.length} />
          <Animated.View style={{
            backgroundColor: 'white',
            height: this.state.bodyHeight,
          }}>
          {this._renderSpinner()}
          <View style={[styles.answersContainer]} >
            {
              answers.map((answer: Object, key: number): any => 
                this._renderAnswer(answer, key)
              )
            }
          </View>
          <AddAnswer login={() => { this.props.login(); }}
            token={this.props.token}
            addAnswer={(answer): Promise<Object> => this.addAnswer(answer)} />
          </Animated.View>
        </ScrollView>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {},
  answersContainer: {
    backgroundColor: 'white',
    marginBottom: 20,
  },
  answerContainer: {},
  row: {
    flexDirection: 'row',
  },
  answer: {
    marginTop: 4,
    marginLeft: 2,
    marginRight: 5,
    flexGrow: 1,
    flex: 1,
  },
  answerText: {
    fontSize: 16,
    color: 'black',
    flexGrow: 1,
  },
  answerFooter: {
    marginTop: 4,
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  username: {
    color: '#BDBDBD',
    fontSize: 15,
  },
  date: {
    color: '#BDBDBD',
  },
  separator: {
    width,
    height: 1,
    backgroundColor: '#f0f0f0',
  },
  addAnswer: {
    width: 145,
    alignSelf: 'center',
    height: 50,
    backgroundColor: 'blue',
    justifyContent: 'center',
    marginBottom: 30,
  },
  addAnswerLabel: {
    textAlign: 'center',
    fontSize: 20,
  },
  progress: {},
  spinner: {
    alignSelf: 'center',
    marginTop: 50,
    marginBottom: 25,
  },
});


const AnswerFragment = {
  answer: gql`
    fragment AnswerData on Answer {
     _id
        answer,
        postedBy {
          username
        }
        date
        rating
        userRating(token: $token) {
          rating
        }
    }
  `,
};

const TaskDetailsQuery = gql`
  query TaskDetails($id: ID!, $token: String) {
    task(id: $id) {
      _id,
      title,
      date,,
      postedBy {
        username
      }
      answers {
        ...AnswerData
      }
    }
  }
  ${AnswerFragment.answer}
`;

const AddAnswerMutation = gql`
  mutation answer($taskId: ID!, $answer: AnswerInput!, $token: String!) {
    answer(taskId: $taskId, answer: $answer, token: $token) {
      ...AnswerData
    }
  }
  ${AnswerFragment.answer}
`;

const RateMutation = gql`
  mutation rating($taskId: ID!, $answerId: ID!, $ratingValue: RatingInput!, $token: String!) {
    rating(taskId: $taskId, answerId: $answerId, ratingValue: $ratingValue, token: $token) {
      rating
    }
  }
`;

function mapStateToProps(state): Object {
  return {
    ...state.user,
  };
}
function mapDispatchToProps(dispatch): Object {
  return {
    login: () => { dispatch(login()); },
  };
}

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  graphql(AddAnswerMutation, {
    props: ({ownProps, mutate}) => ({
        addAnswer: (taskId, answer, token) => mutate({
          variables: {taskId, answer, token},
          optimisticResponse: {
            __typename: 'Mutation',
            answer: {
              __typename: 'Answer',
              _id: null,
              answer: answer,
              date: new Date(),
              postedBy: {
                username: '',
              },
              userRating: null,
              rating: 0,
            },
          },
          updateQueries: {
            TaskDetails: (prev, { mutationResult, ...rest }) => {
              const newAnswer = mutationResult.data.answer;
              const next = {
                ...prev,
                task: {
                  ...prev.task,
                  answers: [
                    ...prev.task.answers,
                    {
                      ...newAnswer,
                    },
                  ],
                },
              };
              return next;
              /*
              {
                ...prev,
                task: {
                  ...prev.task,
                  answers: [
                    ...prev.task.answers,
                    newAnswer,
                  ],
                },
              }
              */
            },
            Venue: (prev, { mutationResult }) => {
              const newAnswer = mutationResult.data.answer;
              const taskInVenueIndex = prev.venue.tasks.findIndex(({_id}): boolean => _id === ownProps.task._id);
              if (taskInVenueIndex <= -1) return prev;
              return {
                ...prev,
                venue: {
                  ...prev.venue,
                  tasks: [
                    ...prev.venue.tasks.slice(0, taskInVenueIndex),
                    {
                      ...prev.venue.tasks[taskInVenueIndex],
                      nbAnswers: prev.venue.tasks[taskInVenueIndex].nbAnswers + 1,
                    },
                    ...prev.venue.tasks.slice(taskInVenueIndex + 1),
                  ],
                },
              };
            },
          },
        }),
    }),
  }),
  graphql(TaskDetailsQuery, {
    options: ({ id, token }): Object => ({
      variables: {id, token},
    }),
    props: ({ ownProps, data: { loading, error, task, ...rest } }): Object => {
      return {
        loading,
        task: task || ownProps.task,
        error,
      };
    },
  }),
  graphql(RateMutation, {
    props: ({mutate}) => ({
        rateAnswer: (taskId, answerId, ratingValue, token) =>
          mutate({
            variables: {taskId, answerId, ratingValue, token},
            updateQueries: {
              TaskDetails: (prev, { mutationResult }) => {
                if (mutationResult.error) return prev;
                console.log('================');
                console.log(mutationResult)
                console.log('{{{{{{{{{{}}}}}}}}}}')
                console.log(prev);
                const next = {

                }
                return prev;
              },
            },
          }),
    }),
  }),
)(TaskDetails);
