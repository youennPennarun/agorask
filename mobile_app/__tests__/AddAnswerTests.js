/* global jest, expect, describe, it */

import 'react-native';
import React from 'react';

import {AddAnswer} from '../src/components/views/taskDetails/AddAnswer';

import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

it('renders correctly', () => {
  const wrapper = shallow(
    <AddAnswer />
  );
  expect(shallowToJson(wrapper)).toMatchSnapshot();
});
