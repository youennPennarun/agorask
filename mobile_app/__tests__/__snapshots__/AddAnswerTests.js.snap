exports[`test renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
    }
  }>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail">
    You need to be logged in to add an answer
  </Text>
  <BuiltButton
    backgroundColor="#3F51B5"
    enabled={true}
    maskBorderRadius={2}
    maskColor="rgba(255,255,255,0.15)"
    maskDuration={200}
    maskEnabled={true}
    onPress={[Function]}
    pointerEvents="box-only"
    rippleColor="rgba(255,255,255,0.2)"
    rippleDuration={200}
    rippleLocation="tapLocation"
    shadowAniEnabled={true}
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "padding": 8,
        },
        Object {
          "backgroundColor": "#3F51B5",
        },
        Object {
          "borderRadius": 2,
          "elevation": 4,
          "shadowColor": "black",
          "shadowOffset": Object {
            "height": 0.5,
            "width": 0,
          },
          "shadowOpacity": 0.7,
          "shadowRadius": 1,
        },
        Object {
          "alignSelf": "center",
          "marginVertical": 10,
          "width": 150,
        },
      ]
    }
    text="Login"
    textStyle={
      Array [
        Object {
          "fontSize": 14,
        },
        Object {
          "color": "white",
          "fontWeight": "bold",
        },
      ]
    } />
</View>
`;
